view/log.go

```diff
@@ -149,6 +149,136 @@ func (li *logView) Read(facts origins.Facts) (int, error) {
 	return l, nil
 }
 
+// merger maintains state of a merged stream of facts from multiple domains.
+type merger struct {
+	iterators []origins.Iterator
+	next      []*origins.Fact
+}
+
+func (m *merger) Next() (*origins.Fact, error) {
+	var maxTrn uint64 = 0
+	var maxI int = -1
+
+	// look at the next fact coming out of each of the the available logs,
+	// and choose the fact with the greatest transaction ID.
+	for i := 0; i < len(m.iterators); i++ {
+		if m.next[i] != nil && m.next[i].Transaction > maxTrn {
+			maxTrn = m.next[i].Transaction
+			maxI = i
+		}
+	}
+
+	if maxI == -1 {
+		return nil, io.EOF
+	}
+
+	// push the newest fact out of the merged stream
+	nextFact := m.next[maxI]
+	fact, err := m.iterators[maxI].Next()
+
+	if err != nil {
+		if err == io.EOF {
+			m.next[maxI] = nil
+		} else {
+			return nil, err
+		}
+	} else {
+		m.next[maxI] = fact
+	}
+
+	return nextFact, nil
+}
+
+func (m *merger) init() error {
+	var err error
+
+	for i := 0; i < len(m.iterators); i++ {
+		m.next[i], err = m.iterators[i].Next()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// merge multiple fact streams into one.
+// Note: iterators return facts in the order opposite to that in which
+// they were committed (i.e. newest facts with largest transaction IDs first),
+// and the Merge operation will preserve that.
+func Merge(iterators []origins.Iterator) (*merger, error) {
```

Comment:
I've recently observed that it is more idiomatic (and less confusing) to use the interface type as the return value for private or wrapper structures like these. Also when passing a variable number of the same type, the [variadic syntax](https://gobyexample.com/variadic-functions) is generally more idiomatic.

```go
func Merge(iters ...origins.Iterator) (origins.Iterator, error) {
    ...
}
```

There are few places that I need to change myself (such as returning a `*logView` value). Also, please capitalize "merge" in the documentation for this function.